### PR TITLE
[Merged by Bors] - feat: add `fermatLastTheoremWith_of_fermatLastTheoremWith_coprime`

### PR DESCRIPTION
**Pull request by riccardobrasca**
We add `fermatLastTheoremWith_of_fermatLastTheoremWith_coprime`: To prove Fermat Last Theorem one can assume that `a`, `b` and `c` are coprime.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
The name should probably be changed. Also, I am not completely sure we want the version for `ℕ` or `ℤ`. Does anyone as an opinion? Do we want both?

**Comment by tb65536** [comment]
I think `hn` is false as stated since you can take `a=1`, `b=0`, `c=1`.

**Comment by MichaelStollBayreuth** [comment]
Yes; you need to add the nontriviality condition to `hn`.

As to naturals vs. integers: there could be a general version for unique factorization (semi)rings, which can be specialized to `Nat` (and `Int`).

**Comment by tb65536** [comment]
> Yes; you need to add the nontriviality condition to `hn`.
> 
> As to naturals vs. integers: there could be a general version for unique factorization (semi)rings, which can be specialized to `Nat` (and `Int`).

I guess you could first prove this theorem under general conditions for `FermatLastTheoremWith`, and then specialize to the natural numbers to deduce the result for `FermatLastTheoremFor`.

**Comment by riccardobrasca** [comment]
> I think `hn` is false as stated since you can take `a=1`, `b=0`, `c=1`.

Ops, well spotted!

**Comment by riccardobrasca** [comment]
> Yes; you need to add the nontriviality condition to `hn`.
> 
> As to naturals vs. integers: there could be a general version for unique factorization (semi)rings, which can be specialized to `Nat` (and `Int`).

[Finset.gcd](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GCDMonoid/Finset.html#Finset.gcd) wants a `NormalizedGCDMonoid`, and in practice this holds very rarely (I don't think we have a reasonable way of getting a semiring with this property, UFD is not enough). I can state the condition by hand though, let me see.

**Comment by riccardobrasca** [comment]
> > > Yes; you need to add the nontriviality condition to `hn`.
> > > As to naturals vs. integers: there could be a general version for unique factorization (semi)rings, which can be specialized to `Nat` (and `Int`).
> > 
> > 
> > [Finset.gcd](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GCDMonoid/Finset.html#Finset.gcd) wants a `NormalizedGCDMonoid`, and in practice this holds very rarely (I don't think we have a reasonable way of getting a semiring with this property, UFD is not enough). I can state the condition by hand though, let me see.
> 
> Well, it seems to me like a waist of time: what we should to is to remove the `NormalizedGCDMonoid` assumption from the results used here.

Now I remember that I have already thought about this: it is not completely trivial, since in reality the `GCD` is only defined up to multiplication by a unit, and in particular it is not really commutative. Annoying...

**Comment by riccardobrasca** [comment]
Wait, `[CommSemiring R] [IsDomain R] [DecidableEq R] [NormalizedGCDMonoid R]` should be ok, I was confused. Let me see and sorry for the noise.

**Comment by kbuzzard** [comment]
In my private FLT repo I have reduced FLT to (a) FLT for n=3 and (b) non-existence of Frey curve, which contains this result and more as a consequence (I have a,b,c coprime and a=3 mod 4 and b=0 mod 32 or something, whatever one needs to make the curve semistable at 2). But given that we're not going live until April I don't object to this going into mathlib now.

**Comment by riccardobrasca** [comment]
> In my private FLT repo I have reduced FLT to (a) FLT for n=3 and (b) non-existence of Frey curve, which contains this result and more as a consequence (I have a,b,c coprime and a=3 mod 4 and b=0 mod 32 or something, whatever one needs to make the curve semistable at 2). But given that we're not going live until April I don't object to this going into mathlib now.

Don't worry about FLT3, I take responsibility for having it in mathlib before the summer (at least PRed...) but I need this.

**Comment by riccardobrasca** [comment]
I've added the general version for a `NormalizedGCDMonoid`.

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/NumberTheory/FLT/Basic.lean [hunk 1]
```diff
@@ -117,3 +119,61 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
 
 lemma fermatLastTheoremFor_iff_rat {n : ℕ} : FermatLastTheoremFor n ↔ FermatLastTheoremWith ℚ n :=
   (fermatLastTheoremWith_nat_int_rat_tfae n).out 0 2
+
+open Finset in
+/-- To prove Fermat Last Theorem in any semiring that is a `NormalizedGCDMonoid` one can assume
+that the `gcd` of `{a, b, c}` is a unit. -/
+lemma fermatLastTheoremWith_of_FermatLastTheoremWith_coprime {n : ℕ} {R : Type*} [CommSemiring R]
+    [IsDomain R] [DecidableEq R] [NormalizedGCDMonoid R]
+    (hn : ∀ a b c : R, a ≠ 0 → b ≠ 0 → c ≠ 0 → IsUnit (({a, b, c} : Finset R).gcd id) →
```
Comment (tb65536) [inline Mathlib/NumberTheory/FLT/Basic.lean:128]:
```suggestion
    (hn : ∀ a b c : R, a ≠ 0 → b ≠ 0 → c ≠ 0 → ({a, b, c} : Finset R).gcd id = 1 →
```

File: Mathlib/NumberTheory/FLT/Basic.lean [hunk 2]
```diff
@@ -117,3 +119,61 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
...
+      rw [← this, hB, hC, mul_assoc, _root_.gcd_mul_left, H,
+        normUnit_mul hdzero fun h ↦ hc <| by simp [h, hC]]
+      congr 3
+      simp only [Units.val_mul, ne_eq, Units.ne_zero, not_false_eq_true, mul_eq_right₀,
+        Units.val_eq_one]
+      nth_rewrite 2 [← mul_one (s.gcd id)] at H
+      exact Units.ext <| by simpa using mul_left_cancel₀ hdzero H
+    · simp [hc]
+  · rw [hA, hB, hC, mul_pow, mul_pow, mul_pow, ← mul_add] at habc
+    simpa [hdn] using habc
```
Comment (tb65536) [inline Mathlib/NumberTheory/FLT/Basic.lean:158]:
```suggestion
  simp only [hA, hB, hC, mul_ne_zero_iff, mul_pow] at ha hb hc habc
  rw [← mul_add, mul_right_inj' (pow_ne_zero n ha.1)] at habc
  refine hn A B C ha.2 hb.2 hc.2 ?_ habc
  rw [← Finset.normalize_gcd, normalize_eq_one]
  obtain ⟨u, hu⟩ := normalize_associated d
  refine ⟨u, mul_left_cancel₀ (mt normalize_eq_zero.mp ha.1) (hu.symm ▸ ?_)⟩
  rw [← Finset.gcd_mul_left, gcd_eq_gcd_image, image_insert, image_insert, image_singleton,
      id_eq, id_eq, id_eq, ← hA, ← hB, ← hC]
```

File: Mathlib/NumberTheory/FLT/Basic.lean [hunk 3]
```diff
@@ -117,3 +119,61 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
 
 lemma fermatLastTheoremFor_iff_rat {n : ℕ} : FermatLastTheoremFor n ↔ FermatLastTheoremWith ℚ n :=
   (fermatLastTheoremWith_nat_int_rat_tfae n).out 0 2
+
+open Finset in
+/-- To prove Fermat Last Theorem in any semiring that is a `NormalizedGCDMonoid` one can assume
+that the `gcd` of `{a, b, c}` is a unit. -/
```
Comment (tb65536) [inline Mathlib/NumberTheory/FLT/Basic.lean:125]:
```suggestion
that the `gcd` of `{a, b, c}` is `1`. -/
```

**Review by tb65536 — COMMENTED** [review]
In a `NormalizedGCDMonoid`, a `gcd` is a unit if and only if it equals 1. So you might as well replace the `IsUnit` condition with "equals 1" (which makes the hypothesis `hn` weaker, thereby making the lemma stronger).

**Comment by riccardobrasca** [comment]
> In a `NormalizedGCDMonoid`, a `gcd` is a unit if and only if it equals 1. So you might as well replace the `IsUnit` condition with "equals 1" (which makes the hypothesis `hn` weaker, thereby making the lemma stronger).

Thanks! [normalize_eq_one](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GCDMonoid/Basic.html#normalize_eq_one) should change name BTW

File: Mathlib/NumberTheory/FLT/Basic.lean [hunk 4]
```diff
@@ -117,3 +119,43 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
...
+    (fun a b c ha hb hc H h ↦ hn a b c ha hb hc H h)
+
+/-- To prove Fermat Last Theorem one can assume that `a`, `b` and `c` are coprime, in the sense
+that the gcd of `{a, b, c}` is `1`. -/
+lemma FermatLastTheoremFor_of_FermatLastTheoremFor_coprime {n : ℕ}
+    (hn : ∀ a b c, a ≠ 0 → b ≠ 0 → c ≠ 0 → ({a, b, c} : Finset ℕ).gcd id = 1 →
+      a ^ n + b ^ n ≠ c ^ n) :
+    FermatLastTheoremFor n :=
+  fermatLastTheoremWith_of_FermatLastTheoremWith_coprime (R := ℕ)
+    (fun a b c ha hb hc H h ↦ hn a b c ha hb hc (by simpa using H) h)
```
Comment (tb65536) [inline Mathlib/NumberTheory/FLT/Basic.lean:161]: Is there any point keeping these anymore?
```suggestion
/-- To prove Fermat Last Theorem for `ℤ` one can assume that `a`, `b` and `c` are coprime, in the
sense that the gcd of `{a, b, c}` is `1`. -/
lemma FermatLastTheoremWith_int_of_FermatLastTheoremWith_int_coprime {n : ℕ}
    (hn : ∀ a b c : ℤ, a ≠ 0 → b ≠ 0 → c ≠ 0 → ({a, b, c} : Finset ℤ).gcd id = 1 →
      a ^ n + b ^ n ≠ c ^ n) :
    FermatLastTheoremWith ℤ n :=
  fermatLastTheoremWith_of_FermatLastTheoremWith_coprime hn

/-- To prove Fermat Last Theorem one can assume that `a`, `b` and `c` are coprime, in the sense
that the gcd of `{a, b, c}` is `1`. -/
lemma FermatLastTheoremFor_of_FermatLastTheoremFor_coprime {n : ℕ}
    (hn : ∀ a b c, a ≠ 0 → b ≠ 0 → c ≠ 0 → ({a, b, c} : Finset ℕ).gcd id = 1 →
      a ^ n + b ^ n ≠ c ^ n) :
    FermatLastTheoremFor n :=
  fermatLastTheoremWith_of_FermatLastTheoremWith_coprime hn
```

Comment (riccardobrasca) [reply Mathlib/NumberTheory/FLT/Basic.lean]: Probably not, you're right. Thanks!

File: Mathlib/NumberTheory/FLT/Basic.lean [hunk 5]
```diff
@@ -117,3 +118,25 @@ lemma fermatLastTheoremFor_iff_int {n : ℕ} : FermatLastTheoremFor n ↔ Fermat
 
 lemma fermatLastTheoremFor_iff_rat {n : ℕ} : FermatLastTheoremFor n ↔ FermatLastTheoremWith ℚ n :=
   (fermatLastTheoremWith_nat_int_rat_tfae n).out 0 2
+
+open Finset in
+/-- To prove Fermat Last Theorem in any semiring that is a `NormalizedGCDMonoid` one can assume
+that the `gcd` of `{a, b, c}` is `1`. -/
+lemma fermatLastTheoremWith_of_FermatLastTheoremWith_coprime {n : ℕ} {R : Type*} [CommSemiring R]
```
Comment (tb65536) [inline Mathlib/NumberTheory/FLT/Basic.lean:125]: Should this be lowercase as well?
```suggestion
lemma fermatLastTheoremWith_of_fermatLastTheoremWith_coprime {n : ℕ} {R : Type*} [CommSemiring R]
```

**Review by tb65536 — APPROVED** [review]

**Comment by tb65536** [comment]
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**